server/src/test/java/org/elasticsearch/search/aggregations/InternalAggregationsTests.java

```diff
@@ -0,0 +1,161 @@
+        }
+        List<SiblingPipelineAggregator> topLevelPipelineAggs = null;
+        if (randomBoolean()) {
+            topLevelPipelineAggs = new ArrayList<>();
+            if (randomBoolean()) {
+                topLevelPipelineAggs.add((SiblingPipelineAggregator)new MaxBucketPipelineAggregationBuilder("name1", "bucket1").create());
+            }
+            if (randomBoolean()) {
+                topLevelPipelineAggs.add((SiblingPipelineAggregator)new AvgBucketPipelineAggregationBuilder("name2", "bucket2").create());
+            }
+            if (randomBoolean()) {
+                topLevelPipelineAggs.add((SiblingPipelineAggregator)new SumBucketPipelineAggregationBuilder("name3", "bucket3").create());
+            }
+        }
+
+        Version version = VersionUtils.randomVersion(random());
+
+        InternalAggregations aggregations = new InternalAggregations(aggsList, topLevelPipelineAggs);
+        try (BytesStreamOutput out = new BytesStreamOutput()) {
+            out.setVersion(version);
+            aggregations.writeTo(out);
+            try (StreamInput in = new NamedWriteableAwareStreamInput(StreamInput.wrap(out.bytes().toBytesRef().bytes), registry)) {
+                in.setVersion(version);
+                InternalAggregations deserialized = InternalAggregations.readAggregations(in);
+                assertEquals(aggregations.aggregations, deserialized.aggregations);
+                if (aggregations.getTopLevelPipelineAggregators() == null) {
+                    assertNull(deserialized.getTopLevelPipelineAggregators());
+                } else {
+                    //TODO update version after backport
+                    if (version.onOrAfter(Version.V_8_0_0)) {
+                        assertNotNull(deserialized.getTopLevelPipelineAggregators());
+                        assertEquals(aggregations.getTopLevelPipelineAggregators().size(),
+                            deserialized.getTopLevelPipelineAggregators().size());
+                        for (int i = 0; i < aggregations.getTopLevelPipelineAggregators().size(); i++) {
+                            SiblingPipelineAggregator siblingPipelineAggregator1 = aggregations.getTopLevelPipelineAggregators().get(i);
+                            SiblingPipelineAggregator siblingPipelineAggregator2 = deserialized.getTopLevelPipelineAggregators().get(i);
+                            assertArrayEquals(siblingPipelineAggregator1.bucketsPaths(), siblingPipelineAggregator2.bucketsPaths());
+                            assertEquals(siblingPipelineAggregator1.name(), siblingPipelineAggregator2.name());
+                        }
+                    } else {
+                        assertNull(deserialized.getTopLevelPipelineAggregators());
+                    }
+                }
+            }
+        }
+    }
+
+    //TODO update version and rename after backport
+    public void testSerializationFromPre_8_0_0() throws IOException {
```

Comment:
I understand the intent of this test and I know that we have similar tests elsewhere but I think it should be moved to a rest test or omitted if we are confident that the existing rest tests are enough to test the bwc serialization. This test checks an internal class that we are allowed to change in a minor release (even a patch release) so I don't think we should use a static representation of the serialization that we'll need to change every time we make a modification to the serialization.

Reply:
I think that yaml tests are overkill for this matter, as they are integration tests and take much longer to run. After the backport, this static version is the binary representation of how we serialized the object prior to 6.7.0, which I am pretty sure we will not change. I can add a comment.

Reply:
> I think that yaml tests are overkill for this matter, as they are integration tests and take much longer to run. 

They are overkilled if we write them only to test serialization but since we have some ccs rest tests already it shouldn't be too costly to add one that checks the support for pipeline aggregations. I also agree that we will probably not change the serialization of this class in 6.7.x but my point was more about the general idea of adding serialized bytes from a previous version in a unit test.

Reply:
I plan to do integration tests for this scenario as part of #40038 , I wanted to add coverage there for the the field collapsing bug as well. I prefer the new java test over the yaml ones personally. But our current CCS integration test don't run against multiple versions, while this unit test makes sure that we can read something that was written from e.g. 6.6 compared to simulating that by calling readFrom on master and setting the version to 6.6. Do you see what I mean? Or am I missing something?

Reply:
I understand the intent but I forgot that we don't run the bwc tests in every module, let's leave it like this for now and we can discuss further in #40038